velox/expression/tests/ExprTest.cpp

```diff
@@ -3386,3 +3386,79 @@ TEST_F(ExprTest, cseOverDictionaryOverConstant) {
   assertEqualVectors(expected, result);
   EXPECT_EQ(5, stats.at("plus").numProcessedRows);
 }
+
+TEST_F(ExprTest, cseOverDictionaryAcrossMultipleExpressions) {
+  // This test verifies that CSE across multiple expressions are evaluated
+  // correctly, that is, make sure peeling is done before attempting to re-use
+  // computed results from CSE.
+  auto indices = makeIndices({1, 3});
+  auto input = makeRowVector({
+      wrapInDictionary(
+          indices,
+          makeFlatVector<StringView>({"aa1"_sv, "bb2"_sv, "cc3"_sv, "dd4"_sv})),
+  });
+  // Case 1: Peeled and unpeeled set of rows have overlap. This will ensure the
+  // right pre-computed values are used.
+  // upper(c0) is the CSE here having c0 as a distinct field. Initially its
+  // distinct fields is empty as concat (its parent) will have the same
+  // fields. If during compilation distinct field is not set when it is
+  // identified as a CSE then it will be empty and peeling
+  // will not occur the second time CSE is employed. Here the peeled rows are
+  // {0,1,2,3} and unpeeled are {0,1}. If peeling is performed in the first
+  // encounter, rows to compute will be {_ , 1, _, 3} and in the second
+  // instance if peeling is not performed then rows to computed would be {0,
+  // 1} where row 0 will be computed and 1 will be re-used so row 1 would have
+  // wrong result.
+  {
+    auto exprSet = compileMultiple(
```

Comment:
Any reason not to use evaluateMultipleWithStats method?

Reply:
evaluateMultipleWithStats uses null result vectors before calling eval() which would result in directly copying over the sharedValues vector when moveOrCopyResult() is called at the end of CSE eval. This would then result in hitting a mismatch on the vector size itself while asserting equality with the expected result. However, I want to make sure that the rows are copied over so that I can Assert on the values of the results instead. Therefore, I am initializing a result vector below with the appropriate size to force moveOrCopyResult() to copy the values.

Reply:
> a mismatch on the vector size itself while asserting equality with the expected result

I assume this can be solved by passing in a slice of result vector to assertEqualVectors or passing extra "size" parameter to assertEqualVectors to request comparison only for [0, size) elements.

> I am initializing a result vector below with the appropriate size to force moveOrCopyResult() to copy the values.

It would be nice to add a comment about that. Consider updating evaluateMultipleWithStats to pass an optional 'results' vector to use to avoid copy paste.